fix(session-replay-browser): Optimize function restart logic

jpollock-ampl opened the pull request:
<!---
Thanks for contributing to the Amplitude TypeScript repository! 🎉

Please fill out the following sections to help us quickly review your pull request.
--->

### Summary

This PR optimizes some of the logic around starting/restarting replay capture.  

Previously, the following three actions would cause the replay to start or restart:
- Focus Change
  - This is desired, we want to restart when this happens
- Session ID Changes (via `setSessionId()`)
  - This is desired, we want to restart when this happens
- An event being emitted from the SDK
  - Undesired, a restart is unnecessary in this scenario

This PR stops the replay capture from being restarted when an event is emitted from the SDK.  

How does this fix work?

The primary change is in `recordEvents()`:
```
if (this.recordCancelCallback && !forceRestart) { 
  // Exit Early
}
```

We will check for an existing callback and if the user does **not** pass forceRestart (default is `forceRestart=false`), we will exit early and not re-create the SDK. 

In the case of "focus" or "session ID changes", we pass `forceRestart=true`, to ensure that a new capture is created.


### Verification

Beyond unit tests, I created a test app and installed this version of the SDK.  I was able to observe the behavior of the SDK on a focus event and then when sending 5 events via a button click:


<img width="726" height="103" alt="Screenshot 2025-09-29 at 3 41 46 PM" src="https://github.com/user-attachments/assets/82f4aedb-837e-44d8-a513-74237e37fe8e" />


The first 3 logs represent the "focus" event.  The fourth log represents the button click.  

--- 

Notes:

Given a call to `track()` in your application, it goes through the Amplitude Analytics SDK's pipeline:

For us, that means the plugin execution pipeline.  So it goes through [timeline.ts, which executes the various types of plugins](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/analytics-core/src/timeline.ts#L83):
- before plugins
- enrichment plugins (this is what our plugin is)
- destination plugins

Our plugin (the SR plugin) is an enrichment plugin ([source](https://github.com/amplitude/Amplitude-TypeScript/blob/a1fc5b7/packages/plugin-session-replay-browser/src/session-replay.ts#L21)), so its execute() [method](https://github.com/amplitude/Amplitude-TypeScript/blob/a1fc5b7/packages/plugin-session-replay-browser/src/session-replay.ts#L132) is called for every single event ([source](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/analytics-core/src/timeline.ts#L131))

The execute() [method](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/plugin-session-replay-browser/src/session-replay.ts#L132) calls:

- [evaluateTargetingAndCapture](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/plugin-session-replay-browser/src/session-replay.ts#L165), which will evaluate targeting conditions ([source code for evaluateTargetingAndCapture](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L290))
- It then calls await [this.recordEvents()](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L344) (if not during initialization)
- OR calls void [this.initialize(true)](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L342) during initialization, which [also calls recordEvents()](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L370)

The recordEvents() method:
- Calls [this.getShouldRecord()](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L527) to determine if recording should happen
- If yes, starts the rrweb capture process (including tearing down the existing instance via [stopRecordingEvents()](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/session-replay-browser/src/session-replay.ts#L532))







### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  <!-- Yes or no -->

jpollock-ampl committed:
fix(session-replay-browser): simplify event initialization logic

jpollock-ampl committed:
fix(session-replay): add forceRestart option to recordEvents method

jpollock-ampl committed:
feat(session-replay): enable recordEvents method to accept configuration object

jpollock-ampl committed:
feat(session-replay): add forceRestart param to initialize

jxiwang reviewed (approved):
Nice! Thanks for adding this!

lewgordon-amplitude reviewed (approved):
LGTM!

jpollock-ampl committed:
fix(session-replay): update debug message for existing capture with forceRestart option

jpollock-ampl committed:
fix(session-replay): require recordEvents() params

jpollock-ampl committed:
fix(session-replay): enhance debug message to include sessionId when starting a new capture

jpollock-ampl commented on packages/session-replay-browser/src/session-replay.ts:
```diff
@@ -546,13 +555,22 @@ export class SessionReplay implements AmplitudeSessionReplay {
+
+    // NOTE: If there is already an existing active recording, exit early unless forceRestart is true
+    if (this.recordCancelCallback && !forceRestart) {
+      this.loggerProvider.debug(`A capture is already in progress and forceRestart is false. Exiting.`);
```
Thank you for the feedback - https://github.com/amplitude/Amplitude-TypeScript/pull/1310/commits/c81dab9a6990efbe52cd7fa4ddbe63a125547345

jpollock-ampl commented on packages/session-replay-browser/src/session-replay.ts:
```diff
@@ -546,13 +555,22 @@ export class SessionReplay implements AmplitudeSessionReplay {
 
-  async recordEvents(shouldLogMetadata = true) {
+  async recordEvents(recordEventsConfig?: { shouldLogMetadata?: boolean; forceRestart?: boolean }) {
+    const { shouldLogMetadata = true, forceRestart = true } = recordEventsConfig || {};
```
Great call - https://github.com/amplitude/Amplitude-TypeScript/pull/1310/commits/f75a51292a5cb894504d4f4a65f59b10d0e1ebbb

jpollock-ampl commented on packages/session-replay-browser/src/session-replay.ts:
```diff
@@ -546,13 +555,22 @@ export class SessionReplay implements AmplitudeSessionReplay {
+      this.loggerProvider.debug(`A capture is already in progress and forceRestart is false. Exiting.`);
+      return;
+    }
+    this.loggerProvider.debug(`Starting new capture for session.`);
```
Done! https://github.com/amplitude/Amplitude-TypeScript/pull/1310/commits/c68312f19611c7f5811d1fc1ee82146bc205a89b

lewgordon-amplitude reviewed (approved):
LGTM!

amplitude-sdk-bot committed:
chore(release): publish

 - @amplitude/plugin-session-replay-browser@1.22.18-fixsdk-init.0
 - @amplitude/segment-session-replay-plugin@0.0.0-fixsdk-init.0
 - @amplitude/session-replay-browser@1.29.0-fixsdk-init.0
